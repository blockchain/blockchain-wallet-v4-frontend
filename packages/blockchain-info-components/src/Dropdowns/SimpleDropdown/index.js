import React from 'react'
import PropTypes from 'prop-types'
import onClickOutside from 'react-onclickoutside'

import Dropdown from './template'

class SimpleDropdown extends React.PureComponent {
  constructor (props) {
    super(props)

    this.state = {
      toggled: props.opened,
      selectedItem:
        props.items.filter(item => item.value === props.selectedValue)[0] ||
        props.items[0]
    }
    this.handleClick = this.handleClick.bind(this)
    this.handleCallback = this.handleCallback.bind(this)
  }

  handleClick () {
    this.setState({ toggled: !this.state.toggled })
  }

  handleClickOutside () {
    this.setState({ toggled: false })
  }

  handleCallback (item) {
    this.setState({ toggled: false, selectedItem: item })
    if (this.props.callback) {
      this.props.callback(item)
    }
  }

  render () {
    const { ...rest } = this.props

    return (
      <Dropdown
        {...rest}
        handleClick={this.handleClick}
        handleCallback={this.handleCallback}
        toggled={this.state.toggled}
        selectedItem={this.state.selectedItem}
      />
    )
  }
}

SimpleDropdown.defaultProps = {
  color: 'brand-secondary',
  opened: false,
  selectedValue: 0,
  uppercase: true,
  down: false
}

SimpleDropdown.propTypes = {
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  items: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })
  ),
  callback: PropTypes.func.isRequired,
  opened: PropTypes.bool,
  uppercase: PropTypes.bool,
  down: PropTypes.bool
}

export default onClickOutside(SimpleDropdown)
