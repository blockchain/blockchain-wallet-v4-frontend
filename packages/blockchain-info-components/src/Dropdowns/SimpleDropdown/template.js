import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import { Icon } from '../../Icons'
import { Palette } from '../../Colors'
import { keysIn } from 'ramda'

const Wrapper = styled.div`
  display: inline-flex;
  text-transform: ${props => (props.uppercase ? 'uppercase' : 'none')};
  position: relative;
`
const ButtonContainer = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  width: inherit;

  & > * {
    color: ${props => props.theme[props.color]}!important;
  }
`
const Button = styled.div`
  display: inline;
  font-family: 'Montserrat', Helvetica, sans-serif;
`
const DropdownIcon = styled(Icon)`
  padding-left: 2px;
`
const DropdownList = styled.ul`
  background-clip: padding-box;
  background-color: ${props => props.theme['white']};
  border: 1px solid ${props => props.theme['gray-1']};
  border-radius: 4px;
  bottom: 0px;
  #box-sizing: border-box;
  box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
  display: ${props => (props.toggled ? 'block' : 'none')};
  float: none;
  height: auto;
  width: inherit;
  line-height: 20px;
  list-style-image: none;
  list-style-position: outside;
  list-style-type: none;
  margin: 2px 0px;
  max-height: 400px;
  min-width: 20px;
  overflow: auto;
  padding: 5px 0px;
  position: absolute;
  right: 0;
  ${props =>
    props.down
      ? 'top: 25px; bottom: auto;'
      : 'top: auto; bottom: 25px;'} z-index: 10;
`

const DropdownItem = styled.li`
  color: ${props => props.theme['gray-5']};
  cursor: pointer;
  padding: 3px 20px;
  font-family: 'Montserrat', Helvetica, sans-serif;
  font-size: 14px;
  font-weight: 300;
  text-align: left;
  text-size-adjust: 100%;
  white-space: nowrap;
`

const Dropdown = props => {
  const {
    color,
    down,
    uppercase,
    toggled,
    selectedItem,
    items,
    handleClick,
    handleCallback
  } = props

  return (
    <Wrapper uppercase={uppercase}>
      <DropdownList toggled={toggled} down={down}>
        {items.map((item, index) => {
          return (
            <DropdownItem key={index} onClick={handleCallback.bind(null, item)}>
              {item.text}
            </DropdownItem>
          )
        })}
      </DropdownList>
      <ButtonContainer color={color} onClick={handleClick}>
        <Button>{selectedItem.text}</Button>
        <DropdownIcon name='down-arrow' size='8px' />
      </ButtonContainer>
    </Wrapper>
  )
}

Dropdown.defaultProps = {
  color: 'brand-secondary',
  toggled: false,
  selectedValue: 0,
  uppercase: true,
  down: false
}

Dropdown.propTypes = {
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  items: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })
  ),
  callback: PropTypes.func.isRequired,
  toggled: PropTypes.bool,
  color: PropTypes.oneOf(keysIn(Palette())),
  uppercase: PropTypes.bool,
  down: PropTypes.bool
}

export default Dropdown
