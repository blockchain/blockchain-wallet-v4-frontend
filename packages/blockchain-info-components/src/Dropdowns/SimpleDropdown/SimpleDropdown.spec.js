import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import SimpleDropdown from './template'

describe('SimpleDropdown component', () => {
  it('default renders correctly', () => {
    const items = [{ text: 'test', value: 'value' }]
    const component = shallow(
      <SimpleDropdown
        items={items}
        selectedItem={{ text: 'test' }}
        callback={() => {}}
        handleClick={() => {}}
        handleCallback={() => {}}
      />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
