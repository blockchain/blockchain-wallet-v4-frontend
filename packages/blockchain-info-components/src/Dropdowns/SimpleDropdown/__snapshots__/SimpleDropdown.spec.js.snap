// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleDropdown component default renders correctly 1`] = `
<template__Wrapper
  uppercase={true}
>
  <template__DropdownList
    down={false}
    toggled={false}
  >
    <template__DropdownItem
      key="0"
      onClick={[Function]}
    >
      test
    </template__DropdownItem>
  </template__DropdownList>
  <template__ButtonContainer
    color="brand-secondary"
    onClick={[Function]}
  >
    <template__Button>
      test
    </template__Button>
    <template__DropdownIcon
      name="down-arrow"
      size="8px"
    />
  </template__ButtonContainer>
</template__Wrapper>
`;
